rlpx.md

```diff
@@ -42,118 +60,305 @@ the authenticity of the message by checking whether
+
+Either side may disconnect if authentication of the first framed packet fails.
+
+Handshake messages:
 
 ```text
+auth = auth-size || enc-auth-body
+auth-size = size of enc-auth-body, encoded as a big-endian 16-bit integer
+auth-vsn = 4
+auth-body = [sig, initiator-pubk, initiator-nonce, auth-vsn, ...]
+enc-auth-body = ecies.encrypt(recipient-pubk, auth-body || auth-padding, auth-size)
+auth-padding = arbitrary data
+```
 
-auth -> E(remote-pubk, S(ephemeral-privk, static-shared-secret ^ nonce) || H(ephemeral-pubk) || pubk || nonce || 0x0)
-auth-ack -> E(remote-pubk, remote-ephemeral-pubk || nonce || 0x0)
+```
+ack = ack-size || enc-ack-body
+ack-size = size of enc-ack-body, encoded as a big-endian 16-bit integer
+ack-vsn = 4
+ack-body = [recipient-ephemeral-pubk, recipient-nonce, ack-vsn, ...]
+enc-ack-body = ecies.encrypt(initiator-pubk, ack-body || ack-padding, ack-size)
+ack-padding = arbitrary data
+```
 
+Implementations must ignore any mismatches in `auth-vsn` and `ack-vsn`. Implementations
+must also ignore any additional list elements in `auth-body` and `ack-body`.
+
+Secrets generated following the exchange of handshake messages:
+
+```text
 static-shared-secret = ecdh.agree(privkey, remote-pubk)
+ephemeral-key = ecdh.agree(ephemeral-privkey, remote-ephemeral-pubk)
+shared-secret = keccak256(ephemeral-key || keccak256(nonce || initiator-nonce))
+aes-secret = keccak256(ephemeral-key || shared-secret)
+mac-secret = keccak256(ephemeral-key || aes-secret)
 ```
 
-Values generated following the handshake (see below for steps):
+## Framing
+
+All messages following the initial handshake are framed. A frame carries a single
+encrypted message belonging to a capability.
+
+The purpose of framing is multiplexing multiple capabilites over a single connection.
+Secondarily, as framed messages yield reasonable demarcation points for message
+authentication codes, supporting an encrypted and authenticated stream becomes
+straight-forward. Frames are encrypted and authenticated via key material generated during
+the handshake.
+
+The frame header provides information about the size of the message and the message's
+source capability. Padding is used to prevent buffer starvation, such that frame
+components are byte-aligned to block size of cipher.
 
 ```text
-ephemeral-shared-secret = ecdh.agree(ephemeral-privkey, remote-ephemeral-pubk)
-shared-secret = keccak256(ephemeral-shared-secret || keccak256(nonce || initiator-nonce))
-aes-secret = keccak256(ephemeral-shared-secret || shared-secret)
-# destroy shared-secret
-mac-secret = keccak256(ephemeral-shared-secret || aes-secret)
-# destroy ephemeral-shared-secret
+frame = header-ciphertext || header-mac || frame-ciphertext || frame-mac
+header-ciphertext = aes(aes-secret, header)
+header = frame-size || header-data || header-padding
+header-data = [protocol-type, context-id]
+capability-id = integer < 2**16, big endian
+context-id = integer < 2**16, big endian
```

Comment:
I think this description is either a)not quite complete or b) needs a footnote that context-id is no longer used. Originally there was quite a bit of info on how to handle multi-framed messages (chunked messages) and the context-id was intended (I think) as the correlating identifier.  Some implementations I have seen simply ignored this feature and went for single-frame messages only.  I vaguely recall hitting a problem in my own implementation, where I was not sure the best way of determining how to split the message into a chunked message if it went over the frame window.....there was some catch-22 in the original spec.

Reply:
The concurrent transfers feature in the original spec was really shaky. A couple of people implemented it but nobody ever turned it on.

Reply:
Changed it now to declare these fields as 'always zero'.